### ISSUE TITLE
Significant performance degradation of the brush rendering

### ISSUE THREAD

**Issue by ksofiyuk** [napari/napari#6579]
### 🐛 Bug Report

I've tested the latest main branch and found out that since my last optimization (#5732) there has been a noticeable degradation of the rendering performance.

When I run `examples/add_labels.py`, I cannot see anything above 35-40 FPS and the brush feels laggy. Moreover, all 16 cores of my CPU are heavily loaded when I use the brush, as it can be seen in the video below:

https://github.com/napari/napari/assets/5459767/2e39c010-2770-4f73-8778-9c68392175bd

In comparison, when I switch to #5732, it shows stable 60 FPS and almost all the CPU cores are free:

https://github.com/napari/napari/assets/5459767/0485ed09-9779-4b71-bb39-a0b10e69c0d8

I think that it is also related to the big changes in the color mapping and rendering code that were introduced in https://github.com/napari/napari/pull/6439, https://github.com/napari/napari/pull/3308

### 💡 Steps to Reproduce

1. I added the following code to measure FPS to `examples/add_labels.py`:
```
def update_fps(fps):
    """Update fps."""
    viewer.text_overlay.text = f"{fps:1.1f} FPS"

viewer.text_overlay.visible = True
viewer.window._qt_viewer.canvas._scene_canvas.measure_fps(callback=update_fps)
```

### 🌎 Environment

napari: 0.1.dev3200+gbb785f3.d20240108
Platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.29
System: Ubuntu 20.04.6 LTS
Python: 3.8.10 (default, Nov 22 2023, 10:22:35)  [GCC 9.4.0]
Qt: 5.15.2
PyQt5: 5.15.10
NumPy: 1.23.0
SciPy: 1.10.0
Dask: 2023.3.2
VisPy: 0.14.1
magicgui: 0.7.2
superqt: 0.6.1
in-n-out: 0.1.7
app-model: 0.2.4
npe2: 0.7.3

OpenGL:
  - GL version:  4.6.0 NVIDIA 525.147.05
  - MAX_TEXTURE_SIZE: 32768

Screens:
  - screen 1: resolution 2560x1440, scale 1.0
  - screen 2: resolution 1920x1080, scale 1.0

Settings path:
  - /home/konstantin/.config/napari/usr_5c0e77c46be77719bc1c0a5ec9c64b94d2a2e50c/settings.yaml
Plugins:
  - napari: 0.1.dev3200+gbb785f3.d20240108 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.1.10 (2 contributions)

**Comment by Czaki** [comment]
did you installed numba in your environment?

**Comment by ksofiyuk** [comment]
Yes, I have `numba==0.58.1` installed.

**Comment by Czaki** [comment]
Please check using examples code from #6580

**Comment by Czaki** [comment]
Because of GPU limitation, it is impossible to have both accuracy and performance for labels above 16 bits.

**Comment by jni** [comment]
Hmmm. But @Czaki in #5732, there was labels remapping to float using a NumPy array indexing operation. That should be faster than our numba code but not >32x faster? (16 cores x 2x fps improvement) I would not consider this closed by #6580 — using np.uint8 or np.uint16 labels is a workaround for the issue that int32+ labels are now slower than before #5732 even. I would say that this is still a performance bottleneck we want to look at closely.

(Hi @ksofiyuk and thanks for the two reports!)

**Comment by Czaki** [comment]
The CPU monitor is not showing a proper workload. It shows 100% on all cores, but it does not consume the whole CPU time. It is a sequence of spikes. And CPU monitor shows top usage along measure period that is much longer than the spike during paint. 


I even worry that here spawning threads generate problems and maybe for small arrays we should use a sequential version. 

I think that it may be worth waiting on sckit-image release, but I may check the sequential version for small arrays. 
Especially in painting, we change every updated pixel to the same value.

**Comment by jni** [comment]
Yup, good ideas.

So would you agree with me to keep this open after #6580?

**Comment by ksofiyuk** [comment]
> The CPU monitor is not showing a proper workload. It shows 100% on all cores, but it does not consume the whole CPU time

This is not correct, the CPU utilization is reported as the queue load of the CPU - when it shows something around 100%, it basically means that there are no free slots to execute other tasks. Moreover, htop averages it over some period of time (if I'm not mistaken, it averages it over the update interval). So, when it shows 100%, it means that the CPU core is fully busy for a whole second, not just for spurious spike moments. Another evidence of this fact - when I use brush for 20-30 second without interruption, the temperature of the CPU starts to rise quickly leading to the fans to turn on to full speed, which means that it is not just an artifact of the load reporting.

> Because of GPU limitation, it is impossible to have both accuracy and performance for labels above 16 bits.
How is it possible that "the GPU limitation" leads to almost 16 CPU cores fully loaded while the previous implementation that uses a single thread and the CPU for the color mapping could process int32 labels without issues?

**Comment by Czaki** [comment]
> How is it possible that "the GPU limitation" leads to almost 16 CPU cores fully loaded while the previous implementation that uses a single thread and the CPU for the color mapping could process int32 labels without issues? In other words, if it is a limitation of the GPU, how does the CPU load have to do with it?

The limitations of GPU is that we could only send directly the 8 and 16 bits data. If input data is above 16bits we need to transform it. Transforming to float32 leads to lost precision above 2**23 and using big float32 textures makes performance degradation (see https://forum.image.sc/t/request-for-testing-napari-0-4-19-release-candidate/87972/2?u=gbokota).

So we remap values to try to fit in uint8 or uint16. But current code always uses multicore, even when work on few pixels (100 in your example) that is our mistake. Most of computational time here is spawning and ending threads.

**Comment by ksofiyuk** [comment]
Taking a moment to step away from the main discussion, isn't it interesting how technology has evolved over the years? That's funny to observe in 2024, that's a simple brush tool without bells and whistles needs to use a GPU and in some circumstances can fully load a modern CPU with 16 cores. I remember using the brush in Microsoft Paint on Windows 98 over two decades ago. Back then, it worked well without major issues on CPUs that are 2-3 orders of magnitude slower than the modern ones. Photoshop with a much more sophisticated brush tool (with different brush shapes, anti-aliasing, the support of transparency gradients and so on) worked well on the old CPUs 20 years ago as well. 

Today we have modern hardware, GPU shaders, jit with numba, multi-threading, but the simplest brush may still feel laggy. How times have changed.

**Comment by Czaki** [comment]
In paint or photoshop you write directly to bitmap. In napari you wrote labels number that need to be transformed to color. 

The source of the problem is that we optimize it for big arrays (where spawn thread is cheap in the context of whole processing), without checking for small data (maybe you could provide some good ideas for benchmarks for this). Optimization for big arrays is required, for example when switching layer of size 1024x1024.

And of course, it could be speedup. But it requires custom fallback to C++ code that is not accepted currently because of lack of c++ (and python glue with c++) knowledge on a group of core developers. A single person knowing this will introduce a single point of failure (Bus factor 1).

**Comment by Czaki** [comment]
@ksofiyuk could you check #6583 (today I got maximum 40 fps even when just moving mouse with brush so cannot check)

**Comment by ksofiyuk** [comment]
> In paint or photoshop you write directly to bitmap. In napari you wrote labels number that need to be transformed to color.

Displayed color is 3 8-bit integers, do I understand correctly that transforming, for example from int32 to 3 uint8, makes the task 100-1000 times computationally more expensive? Or maybe writing 32-bit integers to an array is orders of magnitudes more expensive than writing to an array of uint8?

Moreover, when we are talking about rendering, we should care only about the texture that needs to be sent to a GPU and displayed on your monitor, which is essentially the same bitmap as in Photoshop or Paint. As far as I understand, the current performance bottleneck is on the side of the rendering and texture transformations, not on the side of updating label arrays itself. And the process of updating a texture for rendering doesn't differ in any sense from what is happening when you draw in Paint. How label arrays are internally represented should not affect the rendering performance if everything is implemented correctly.

**Comment by ksofiyuk** [comment]
> @ksofiyuk could you check #6583 (today I got maximum 40 fps even when just moving mouse with brush so cannot check)

It has the same problem on bigger images (with 1000 more pixels width). I tested it with an image of (1624, 2272), it shows 35 FPS max while loading all my CPU cores almost to 100%. The code from #5732 works smoothly on that image and shows stable 60 FPS.

You can test it on a large resolution by tiling images from `add_labels.py`:
```python
import numpy as np
image = np.tile(image, (8, 8))
label_image = np.tile(label_image, (8, 8))

# initialise viewer with coins image
viewer = napari.view_image(image, name='coins', rgb=False)

# add the labels
label_layer = viewer.add_labels(label_image, name='segmentation')
```

It looks like all this multi-threading only slows everything down even on larger images.

**Comment by Czaki** [comment]
> Displayed color is 3 8-bit integers, do I understand correctly that transforming, for example from int32 to 3 uint8, makes the task 100-1000 times computationally more expensive? Or maybe writing 32-bit integers to an array is orders of magnitudes more expensive than writing to an array of uint8?

At first. Modulo operation is expensive. And auto-color mode is a sequence of colors, so if value is above the length of this sequence, there is a need to use modulo operation to calculate proper color. 
At second, the expensive part is spawn of thread. That is omitted in #6583

> As far as I understand, the current performance bottleneck is on the side of the rendering and texture transformations

Texture transform before send to GPU

> And the process of updating a texture for rendering doesn't differ in any sense from what is happening when you draw in Paint

Until you use contours, yes, it is possible to use precomputed value in case of painting in non-contour mode. 




> It has the same problem on bigger images (with 1000 more pixels width). I tested it with an image of (1624, 2272), it shows 35 FPS max while loading all my CPU cores almost to 100%. The code from #5732 works smoothly on that image and shows stable 60 FPS.

It looks like you wrongly checkout code or use huge size of brush. As in this PR if the size of updated area is below 1000 pixels, then it should use sequential code.

**Comment by ksofiyuk** [comment]
> It looks like you wrongly checkout code or use huge size of brush. As in this PR if the size of updated area is below 1000 pixels, then it should use sequential code.

No, I double-checked it, I use the standard brush size of 10 to measure the performance.

When the image is larger, the slowdown is even more noticeable. If you open an image with the resolution of `(6496, 9088)` (32x32 tile),  #6583 shows only 15 FPS on my computer, #5732 on the other hand shows 45 FPS on the same setup.

The thing is that the sequential code seems to always work faster on my machine even on larger images, let alone not using all my 16 cores. When I change your constant for sequential code to 10000, then in the example above, the FPS rises to 30, but still it is slower than it is with #5732

**Comment by Czaki** [comment]
Modulo will always be slower than just cast to float. Unfortunately, this is how computers works. 

And I do not understand how it may happen that a brush of size 10 (100 pixels square) could not hit in condition added in #6583. 

I have updated #6583 to use cached mapping.

**Comment by ksofiyuk** [comment]
Here are the videos that demonstrate it:

1. #6583 on a large image ((6496, 9088), 32x32 tile). It shows 15-20 FPS and almost all the CPU cores are busy:

https://github.com/napari/napari/assets/5459767/c2b7e955-b28f-4f25-9bca-e1c85e5943d2



2. #5732 on the same image. It shows 40-50 FPS and uses only 2-3 CPU cores:


https://github.com/napari/napari/assets/5459767/5ae740bf-b9b8-4178-a2af-e03e88b68cce

**Comment by Czaki** [comment]
and with latest change?

**Comment by ksofiyuk** [comment]
> I have updated #6583 to use cached mapping.

After this, it starts to work really fast (60 FPS on the large image), however, there seems to be a bug, it produces the following rectangular artifacts during drawing:
![image](https://github.com/napari/napari/assets/5459767/f3e3c254-8ebe-405e-83bd-f59de0b5890e)

**Comment by Czaki** [comment]
I see. Something wrongs with cache.

**Comment by Czaki** [comment]
Looks fixed. I will put a longer description in PR description.

**Comment by Czaki** [comment]
(there is one problem, starting point is not properly painted, working to fix)

**Comment by ksofiyuk** [comment]
The previous bug seems to be fixed, however, with the fix the performance is not so great (20-35 FPS on the large image).

**Comment by Czaki** [comment]
Seething is strange, as in the meantime I only changed the cache setup. So I do not have idea how it could change fps from 60 to 45. 

This change even limits a number of pixels to update.

**Comment by jni** [comment]
I just want to point to some historical resources here, because some of the discussion above confuses what was going on at different stages of development.

#5732 is pre-#3308. This means that it was not just a cast to float that was happening — in the first step, integers were being spread into floats in [0, 1] using [`low_discrepancy_image`](https://github.com/napari/napari/blob/3868f5c8c6f4cbd5704b35ed0fa9bccd19be3435/napari/utils/colormaps/colormap_utils.py#L235-L262). For label values smaller than the data size, we [created an `_all_vals` array](https://github.com/napari/napari/blob/3868f5c8c6f4cbd5704b35ed0fa9bccd19be3435/napari/layers/labels/labels.py#L863-L905) that stored the mapping for each label. This then [used NumPy array indexing](https://github.com/napari/napari/blob/3868f5c8c6f4cbd5704b35ed0fa9bccd19be3435/napari/layers/labels/labels.py#L847) to map the integers to the float values needed by the texture. As discussed in scikit-image/scikit-image#7262, this is much faster than dictionary lookup — the prange solution merely throws more cores at the problem, but does not change the fundamentals of the operation.

#3308 changed all that by simply casting the data as float32 and sending it directly to the GPU. This was *much* faster when converting a large amount of data (e.g. when computing a 3D slice from a 4D+ array), and allowed us to do accurate color mapping up to `2**23` colors, which closed a *bunch* of bugs.

Unfortunately *after* slicing, the fps when displaying the float32 texture was *terrible*. This appears to be an OpenGL limitation (or at least an issue with how VisPy's OpenGL shaders work). We fixed that in #6439 and subsequent PRs, but we wanted to avoid bringing back the issues that #3308 resolved, so the resulting mapping was slower. I hoped that with the caching strategy from #5732 it wouldn't matter, but clearly that was wrong.

@ksofiyuk MS Paint had a much simpler set of constraints than napari does. The MS engineers did not have to worry about drawing into 64-bit numeric arrays with potentially remote data. The color texture *was* the data. I don't think the comparison is helpful in this discussion.

> "do I understand correctly that transforming, for example from int32 to 3 uint8, makes the task 100-1000 times computationally more expensive?"

Well, it can certainly be more expensive given certain constraints, such as: 
- the user should be able to determine the allowable set of colors that the integers can be mapped to
- the user can give arrays of any integer dtype, and expects the mapping to be consistent between dtypes
- the user can specify a color mapping by hand using a dictionary
- the mapping function should be numerically stable so that it will give the same results in the CPU and GPU
- the data update and display need to be kept in sync in a broad range of situations (see e.g. #6079 for a bug that was caused by #5732.)

Maybe that doesn't account for 100-1000, but it plausibly accounts for 10-100. As @Czaki has hinted, we have barely begun to explore the optimization options available.

If you look at #3308, you'll see that it closes a large number of reported bugs. Unfortunately, even though it sped up slicing and painting, it also resulted in a bunch of performance regressions, which we are still untangling. We really appreciate patience and any help you can provide as we work through it! But saying "you are slower than MS Paint 20 years ago" is *not* helpful, nor motivating. (Well, except for @Czaki who I must admit looks pretty motivated right now! 😂)

I must say I'm not loving the direction in #6583 and am slightly more inclined to try bringing back some kind of array-based mapping for small label values, like described above in pre-#3308.

btw, just to clarify something: @ksofiyuk if you use np.uint8 or np.uint16 labels, you get fast performance, right?
